### ISSUE TITLE
Feature Request: ErrorResponse needs a busy flag/indicator

### ISSUE THREAD

**Issue by haneefdm** [microsoft/debug-adapter-protocol#323]
## Introduction
We have a generic ErrorResponse interface. I propose that an enhancement is needed to clarify one common problem.

Generally, there are two classes of problems

* The request failed because of a variety of reasons and this is infinite
* The request failed because the adapter or the debugger (gdb, lldb, etc.) does not allow certain commands when the program is in the running stage.

I propose that we have a flag in `ErrorResponse.body.isBusy` interface to say that the failure is because the program/adapter/debugger is currently running. We can also add it to the `Message` interface but this is used generically in other places and might seem out of place -- unless documented. Six or half a dozen.

## Rationale:

We have 3rd party extensions (not part of the DA extensions) that work with multiple adapters. We all have to make requests from DAs to function. Memory Viewers, RTOS viewers, Peripheral Viewers, etc. They all rely on at least two things.

1. Evaluate Expressions
2. Memory reads and writes

We make these requests when we know (via a debug-tracker or custom-event) to know when the program pauses. The problem is by the time we make that request, the program is in running state and returns an error. We have no idea if the failure is because of of a `busy` problem or something else. In most cases, we can recover from it and retry when the program pauses next time. But, if it is not because of `busy` status, we should never try that again to create minimal disturbance in the Debug Console or otherwise.

When we had our own client and DA, we had a mechanism to say that. But now, we want our client to work with any debugger and refactor these features into separate extensions so more people can use them.

We now have various MS tools (Embedded Tools), Cortex-Debug, memview, RTOS viewers, wanting to support a variety of debuggers (cppdbg, cortex-debug, cspy, any debugger than can support the memory operations). If we had a consistent way of specifying this common `reason` for failure, it can benefit everyone.

We have an RTOS detector (both mine and MS Embedded Tools), we have to do multiple `evaluate`s to see if an RTOS is active. We run into errors before we are done with our first set of queries to validate a set of RTOSes and we declare that RTOS is not visible and won't try again. From the user's point of view, it is clearly there and our extension is useless. If we know that the problem was due to the program being `busy` then we can/should retry until we know for sure it will never work and notify users one way or another. We do recover sometimes based on observing how each individual adapter behaves when the program is busy and hard coding that. This is not scaleable or portable or survives evolving adapters.

## Documentation
Clients may use this flag if they can take advantage of it but it will be no different from any other error. It is just an additional indication of why an error occurred. They can use it to try to recover/retry.

**Comment by haneefdm** [comment]
@benmcmorran you may want to comment and suggest edits to the proposal/rationale
@connor4312 tagging you in case you want to do the same and since we [discussed this](https://github.com/microsoft/vscode/issues/155597)

**Comment by puremourning** [comment]
Would it be more general for the adapter to say 'can retry' or 'no retry' ? Busy might be one of a number of reasons perhaps for transient failure.

**Comment by haneefdm** [comment]
Good point about other reasons. But to me `can retry` is also ambiguous if other requests will fail or just this one. In many cases, all future releases will fail until it is not busy. So to me `busy` means other requests that require a program to in stopped stat will also fail. Image this

But I am open to either. If we use multiple options it can be  `'can-retry' | 'never-retry' | undefined` and it is more extensible for other use cases. I don't see a good use for `never-retry` right now but it does not hurt.

Maybe `can-retry` would be `busy` I really want to know if the debuger is in running status is why it failed.

**Comment by puremourning** [comment]
Sure. Was just a thought. I don't have a strong feeling about it.

**Comment by benmcmorran** [comment]
At least for the Embedded Tools extension, a specific `busy` status would be useful to ensure we're providing the correct actionable guidance to the user. Right now, we have to show this generic error message for any failure, which may or may not actually help the user resolve the issue.

> An error occurred while loading details. Make sure the debugger is paused and try again.

Related to this, perhaps the DAP spec also needs to be clearer about when an error response is required. For example, I've seen inconsistencies in DA responses to `evaluate` requests for undefined variables. cppdbg returns an error message in the `result` of the `EvaluateResponse` and sets `variablesReference` to zero, while IAR's DA returns an `ErrorResponse`.

**Comment by haneefdm** [comment]
I think the evaluate reqest is much more complicated. We (cortex-debug) return either an error response or `null`normal response depending on the `context`. It is based on observing what Watch window does, vs hover vs repl, we modified our behavior. Was not a good experience

**Comment by haneefdm** [comment]
Thanks, @benmcmorran. A couple of years ago, I only worried about ONE client. VSCode. While knowing that could be other clients like Eclipse Che, etc. I think the definition of "client" has moved on in my head. Given VSCode APIs, we can have multiple clients for a DA concurrently.

I don't think it was ever the intent of the protocol to say a DA has to adapt to the client. But should clients reverse-engineer how the DA responds to error situations? Mostly, I do think the spec deals with the non-error situations really well.

I think some sort of standardization of failure responses is in order. I know, I am probably killing my own specific request by asking to generalize it ... so be it.

I hope @weinand could chime in.

**Comment by roblourens** [comment]
> In most cases, we can recover from it and retry when the program pauses next time. But, if it is not because of busy status, we should never try that again to create minimal disturbance in the Debug Console or otherwise.

Does that mean that if there is any sort of unexpected error from the DA, the functionality of your extension shuts down and it never makes another request to the DA? That isn't really what I would expect. It seems to me like it would be ok to just retry the requests on each pause.

I also am not sure how the `id` field on `Message` is currently used, but it seems like DAs could use this to communicate the type of error, and document that somewhere. I know that we did that for the old node debug extension. All DAs will have different sorts of errors for different contexts, and it seems hard to get them to all use one flag consistently.

LSP declares some error codes like this: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseError I don't see anything that matches what we're talking about here. Some language that I like is in the remote resolver API, errors are either "temporary" or not.

**Comment by haneefdm** [comment]
> Does that mean that if there is any sort of unexpected error from the DA, the functionality of your extension shuts down and it never makes another request to the DA? That isn't really what I would expect. It seems to me like it would be ok to just retry the requests on each pause.

First, we are not talking about an unexpected error. We are talking about something specific that happens all the time and it will be across many DAs, and it is not even a real failure

Because users see our requests in the Debug Console (when debug enabled, we tend to generate a lot of chatter), especially during fast single stepping. For this class of requests, if it fails once, it will fail every time (99.9%). We want to try it once for real (when not busy) and then we never try that again, if there is an error and also notify the user. It is not one query we make. We make a series of queries and we will look bad generating so much chatter repeatedly when we know they will fail.

Also, each time we fail what, do we tell the user? We would like to declare it either works or doesn't. And if busy, we should not be generating errors. Staying silent and just keeping on retrying is not a good thing.

Currently, there is a field
```typescript
interface Response extends ProtocolMessage {
  ...
  /**
   * Contains the raw error in short form if `success` is false.
   * This raw error might be interpreted by the client and is not shown in the
   * UI.
   * Some predefined values exist.
   * Values: 
   * 'cancelled': request was cancelled.
   * etc.
   */
   message?: 'cancelled' | string;
}
```

This is not unlike the `cancelled` where one is initiated by the client and the `busy` is because the DA was busy. I wouldn't mind if the `message` field also had a way of saying `busy`, but I have not thought that through.

>  also am not sure how the id field on Message

AFAIK, `Message.id` is not used by anyone other than for display purposes. Only the debug adapter knows what that id means. A debug adapter could document all its error messages by number for instance. Or we can identify where in our code we generated that message.

Please note that I/we (myself and MS Embedded Tools and others) are looking to work with any DA. We will have to reserve an ID and tell everyone that is reserved. There is no good way to communicate this to others. It is much easier to convince DA authors to use a busy flag. This is why we want the DAP to handle this very common case.

## Actual usage scenario. RTOS discovery and display

- This has to do with detecting RTOSes in an embedded program and providing useful information on the current state of the OS. We list thread statistics, thread names, priorities, current/peak stack usage, %runtime, etc. We can also show which semaphores/queues threads are waiting on.
- Generally, all RTOSes have some `required` global variables and many optional ones. Each RTOS has about a dozen such variables and extensions like ours will try to find these variables (by evaluating)
- We support 5 RTOSes, and [Microsoft Embedded Tools](https://github.com/microsoft/vscode-embedded-tools) currently supports 3. So that is 8, for now. This list is growing.
- At the start of a session, we try to evaluate all the required global variables for each OS. If even a single `required` variable for a given RTOS does not exist, we never try again. This is why the first probe matters
- We also try to evaluate all the optional variables and if they do not exist, we adjust the output and have contingencies for failures in future pauses. We also know that if it ever existed, it will always exist, so, the first probe matters.
- We also end up making memory reads for the stack of each RTOS thread (the number of threads is user-defined).

This may sound like a very archaic way of doing things but this is how it is done in the Embedded world and even in other IDEs. This also applies to Microsoft's own Azure RTOS.

Because the list of RTOSes we support is different from the list of MS Embedded Tools, both extensions run concurrently and do their own probes. Both of us, support multiple DAs. I am sure there will be proprietary ones coming doing their own probes.

**Comment by haneefdm** [comment]
Btw, we do have another use case (memory viewer) where we intended to keep trying every time the debugger pauses. This is because unlike in the RTOS case, they may succeed in the future, and the user has some control over it.  Even then, a `busy` indicator is helpful to the user and the extension for a better user experience.

**Comment by connor4312** [comment]
I think the minimal approach for this would be to extend the `message` of the response with an additional known variant:

```ts
interface Response extends ProtocolMessage {
  /**
   * Contains the raw error in short form if `success` is false.
   * This raw error might be interpreted by the client and is not shown in the
   * UI.
   * Some predefined values exist.
   * Values: 
   * 'cancelled': request was cancelled.
   * 'notStopped': request should be retried once the adapter is in a 'stopped' state.
   * etc.
   */
  message?: 'cancelled' | 'notStopped' | string;

// ...
```

I prefer this for increased specificity over "busy", which could mean any number of things,

**Comment by haneefdm** [comment]
@connor4312 I think that would work. I like `notStopped ` as well. And, yes, it would solve our use case issue.

I am hoping @benmcmorran would also chime in for his use case in the Embedded Tools extension.

This can also make our transition plans for the DAs and clients simpler while staying compatible.

**Comment by benmcmorran** [comment]
That looks good to me for Microsoft's Embedded Tools extension. It should be enough for us to provide more precise error messages to users and know when to retry.

**Comment by roblourens** [comment]
>> Does that mean that if there is any sort of unexpected error from the DA, the functionality of your extension shuts down and it never makes another request to the DA? That isn't really what I would expect. It seems to me like it would be ok to just retry the requests on each pause.
>
> First, we are not talking about an unexpected error. We are talking about something specific that happens all the time and it will be across many DAs, and it is not even a real failure

I was not referring to failures that happen because the DA isn't paused. I was talking about any other kind of error. It sounds like the reason that you need to differentiate them is because you handle them differently.

> But, if it is not because of busy status, we should never try that again to create minimal disturbance in the Debug Console or otherwise.

And that is what I'm referring to and I'm wondering whether the different handling is necessary. I am fine with the proposed `notStopped` message. I was just wondering whether it gives you the right context in all cases. There are other errors that you will get that could be ignored and shouldn't be treated as fatal. But you probably know that and looking at the example you gave, I don't think that's the issue for you. I think `notStopped` is fine but I suppose we should run it past the authors of the relevant DAs to be sure that it can/will be implemented in the way you expect.

**Comment by haneefdm** [comment]
> we should run it past the authors of the relevant DAs to be sure that it can/will be implemented in the way you expect

Sure, I am one of the DA authors, but also a client. Any DA will support evaluations but very few support memory-requests. Memory requests and to some extent disassembly only make sense for programs like C/C++. It is us and `cppdbg` (yes, there are a couple more). `cppdbg` owner is @wardenGnaw and hopefully, he can comment. We are both trying to be compatible with each other and serve clients like Embedded Tools (and my own). And the protocol is the best way.

<img width="583" alt="image" src="https://user-images.githubusercontent.com/41269583/189609522-b62274b0-e30e-49e0-9ae0-6b98e587d035.png">

The first one above is `cppdbg` and the second one is us.

There are a couple of others "Debug by @WebFreak001". There is also lldb, go and a few others I know of, but not listed on the [Implementations Page](https://microsoft.github.io/debug-adapter-protocol/implementors/adapters/)

The good thing is a DA does not HAVE to implement it or do it right away. It will continue to work the way it is right now -- in a non-good way.

> I was just wondering whether it gives you the right context in all cases. There are other errors that you will get that could be ignored and shouldn't be treated as fatal

As a client, we have a pretty good idea of when we should never retry. If a global symbol evaluate fails in a C program, it will never work again -- when there is no delayed so/DLL load situation. Expressions however are dependent on the context (thread/frame). Generally, we know when to retry (most of the time for memory requests) and when never to retry (for global variables)

Regardless, the `notStopped` is always helpful to a client and it doesn't matter what it does with that information.

**Comment by weinand** [comment]
Sorry for not chiming in earlier...

Adding a "busy" flag to an error response does not sound like a big deal.

However, we always should keep in mind that DAP is not a universal API for connecting software tools.
DAP is only a protocol to "connect a generic debug UI with real debuggers through strings".

See the DAP spec's intro:

<img width="753" alt="CleanShot 2022-09-12 at 15 37 33@2x" src="https://user-images.githubusercontent.com/1898161/189669010-fc16d740-258d-4018-a6c3-6a11d5751265.png">

So, we do not want to add features to DAP that are used outside of the "generic debugger UI" domain.

That means when adding new features to DAP we always need to answer the questions "how will a generic debugger UI profit from the feature" or "how are clients expected to surface/implement the feature".

If the feature is not expected to be surfaced in the generic debugger UI, but only used in some other context, I would not want to add the feature to the spec but suggest to use a private feature (flag) instead.

From reading through the discussion from above, I was not able to understand how the "busy" flag would be surfaced in a client like VS Code.

@puremourning made a good step in the right direction to call the flag "can retry" or "no retry" because it makes clear what the client is expected to do.

But may be we should really try to first state the problem (e.g. "support recover/retry of DAP requests") and then come up with ways how to solve it (e.g. "add a busy flag").

**Comment by connor4312** [comment]
> flag "can retry" or "no retry" because it makes clear what the client is expected to do.

I considered this, but don't think it's idea for this scenario. An adapter that supports the proposed flag would want to be told to re-attempt the request as soon as the DA is stopped, versus a generic "retry later" which might end up happening on an interval or debounce.

**Comment by haneefdm** [comment]
Speaking of generic UI, the proposed flag would have been great way for DAs to respond for Hover requests, because they occur at totally random times and at least in my DA, I had to figure out how to respond to Hover messages when there are errors or more importantly when busy (return null or empty string or ErrorResponse or what) until we figured out what will work.

Quite possible, I missed some documentation on how to deal with such error-like situations for hover.

**Comment by WardenGnaw** [comment]
> cppdbg owner is @WardenGnaw and hopefully, he can comment. 

This sounds fine for `cppdbg` since we just return a ErrorResponse with the message saying the debuggee is not stopped.
We can use this proposed flag when we know the target is not stopped. 

https://github.com/microsoft/MIEngine/blob/7ece35bcb707731c92bb4fdaf76284c8e2b5fb47/src/OpenDebugAD7/AD7DebugSession.cs#L2947

+@gregg-miskelly @andrewcrawley for any additional comments.

**Comment by haneefdm** [comment]
> This sounds fine for cppdbg since we just return a ErrorResponse with the message saying the debuggee is not stopped.

Thanks, @WardenGnaw, we plan to do that exact thing.
